File: Sources/MacOnScreenChat/EnvLoader.swift
import Foundation

/// Reads secrets (e.g. GEMINI_API_KEY) from the process environment or a local
/// `.env` file — never bundled, never committed. Lookup order:
///   1. process environment
///   2. ~/Library/Application Support/MacOnScreenChat/.env   (used by the .app)
///   3. ./.env in the current directory                      (used by `swift run`)
enum EnvLoader {
    static func value(_ key: String) -> String? {
        if let v = ProcessInfo.processInfo.environment[key], !v.isEmpty {
            return v
        }
        for url in candidateFiles() {
            if let v = parse(url)[key], !v.isEmpty {
                return v
            }
        }
        return nil
    }

    private static func candidateFiles() -> [URL] {
        var urls: [URL] = []
        if let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
            urls.append(base.appendingPathComponent("MacOnScreenChat/.env"))
        }
        urls.append(URL(fileURLWithPath: FileManager.default.currentDirectoryPath).appendingPathComponent(".env"))
        return urls
    }

    private static func parse(_ url: URL) -> [String: String] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [:] }
        var result: [String: String] = [:]
        for rawLine in text.split(whereSeparator: { $0 == "\n" || $0 == "\r" }) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty, !line.hasPrefix("#"), let eq = line.firstIndex(of: "=") else { continue }
            let key = line[..<eq].trimmingCharacters(in: .whitespaces)
            var val = line[line.index(after: eq)...].trimmingCharacters(in: .whitespaces)
            if val.count >= 2, val.hasPrefix("\""), val.hasSuffix("\"") {
                val = String(val.dropFirst().dropLast())
            }
            result[key] = val
        }
        return result
    }
}

File: windows/src/OnScreenChat.Core/Config/EnvLoader.cs
namespace OnScreenChat.Core.Config;

/// <summary>
/// Reads secrets (e.g. GEMINI_API_KEY) from the process environment or a local
/// <c>.env</c> file — never bundled, never committed. Lookup order:
/// <list type="number">
///   <item>process environment</item>
///   <item><c>%LOCALAPPDATA%\OnScreenChat\.env</c> (used by the installed app)</item>
///   <item><c>.\.env</c> in the current directory (used by <c>dotnet run</c>)</item>
/// </list>
/// </summary>
public static class EnvLoader
{
    public static string? Value(string key)
    {
        var fromProcess = Environment.GetEnvironmentVariable(key);
        if (!string.IsNullOrEmpty(fromProcess)) return fromProcess;

        foreach (var file in CandidateFiles())
        {
            if (Parse(file).TryGetValue(key, out var value) && !string.IsNullOrEmpty(value))
            {
                return value;
            }
        }

        return null;
    }

    private static IEnumerable<string> CandidateFiles()
    {
        yield return AppPaths.EnvFile;
        yield return Path.Combine(Directory.GetCurrentDirectory(), ".env");
    }

    public static IReadOnlyDictionary<string, string> Parse(string path)
    {
        if (!File.Exists(path)) return new Dictionary<string, string>();

        string text;
        try
        {
            text = File.ReadAllText(path);
        }
        catch (IOException)
        {
            return new Dictionary<string, string>();
        }

        return ParseText(text);
    }

    public static IReadOnlyDictionary<string, string> ParseText(string text)
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);

        foreach (var rawLine in text.Split(['\n', '\r'], StringSplitOptions.RemoveEmptyEntries))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#')) continue;

            var separator = line.IndexOf('=');
            if (separator < 0) continue;

            var key = line[..separator].Trim();
            var value = line[(separator + 1)..].Trim();

            if (value.Length >= 2 && value.StartsWith('"') && value.EndsWith('"'))
            {
                value = value[1..^1];
            }

            result[key] = value;
        }

        return result;
    }
}
